File: String sol/Longest substring without repeation/longest substring without repeation.cpp

// Given a string s, find the length of the longest substring without duplicate characters.

// optimal approach.

#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> last(256, -1);  // hash array for ASCII chars
        int left = 0, maxLen = 0;

        for (int right = 0; right < s.size(); ++right) {
            unsigned char c = s[right];
            if (last[c] >= left) {
                // duplicate found, shift left
                left = last[c] + 1;
            }
            last[c] = right;  // update last seen index
            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};

File: String_sol/Longest substring without repeation/longest substring without repeation.py

# Given a string s, find the length of the longest substring without duplicate characters.

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last = [-1] * 256   # last seen index for ASCII chars
        left = 0
        maxLen = 0

        for right, ch in enumerate(s):
            c = ord(ch)     # convert char to ASCII index
            if last[c] >= left:
                # duplicate found, move left pointer
                left = last[c] + 1
            last[c] = right  # update last seen index
            maxLen = max(maxLen, right - left + 1)

        return maxLen
